discourse.Dockerfile

```diff
@@ -29,97 +24,92 @@ ENV CONTAINER_APP_GID ${CONTAINER_APP_GID:-200}
-# Remove additions incompatible with Discourse versions < 2.8
-RUN cd ${CONTAINER_APP_ROOT}/app/plugins/discourse-saml && git reset --hard 851f6cebe3fdd48019660b236a447abb6ddf9c89
-RUN git clone https://github.com/discourse/discourse-solved.git ${CONTAINER_APP_ROOT}/app/plugins/discourse-solved
-RUN git clone https://github.com/canonical-web-and-design/discourse-markdown-note.git ${CONTAINER_APP_ROOT}/app/plugins/discourse-markdown-note
-RUN git clone https://github.com/unfoldingWord-dev/discourse-mermaid.git ${CONTAINER_APP_ROOT}/app/plugins/discourse-mermaid
-RUN chown -R ${CONTAINER_APP_USERNAME}:${CONTAINER_APP_GROUP} ${CONTAINER_APP_ROOT}/app/plugins
+    && echo 'source "https://rubygems.org"' > "${PLUGINS_DIR}/discourse-saml/Gemfile" \
+    && grep -e ^gem "${PLUGINS_DIR}/discourse-saml/plugin.rb" >> "${PLUGINS_DIR}/discourse-saml/Gemfile" \
+    && git clone https://github.com/discourse/discourse-solved.git "${PLUGINS_DIR}/discourse-solved" \
+    && git clone https://github.com/canonical-web-and-design/discourse-markdown-note.git "${PLUGINS_DIR}/discourse-markdown-note" \
+    && git clone https://github.com/unfoldingWord-dev/discourse-mermaid.git "${PLUGINS_DIR}/discourse-mermaid" \
+    && chown -R "${CONTAINER_APP_USERNAME}:${CONTAINER_APP_GROUP}" "${PLUGINS_DIR}" \
 # Have to determine the gems needed and install them now, otherwise Discourse will
 # try to install them at runtime, which may not work due to network access issues.
-RUN cd ${CONTAINER_APP_ROOT}/app/plugins/discourse-saml && install -d -o ${CONTAINER_APP_USERNAME}  gems && echo 'source "https://rubygems.org"' > Gemfile ; grep -e ^gem plugin.rb >> Gemfile && su -s /bin/bash -c 'gem install bundler' ${CONTAINER_APP_USERNAME} && su -s /bin/bash -c '${CONTAINER_APP_ROOT}/app/bin/bundle install --gemfile=Gemfile --path=${CONTAINER_APP_ROOT}/app/plugins/discourse-saml/gems' ${CONTAINER_APP_USERNAME} && cd gems && ln -s ruby/* ./
-RUN cd ${CONTAINER_APP_ROOT}/app && su -s /bin/bash -c 'bin/bundle install' ${CONTAINER_APP_USERNAME}
-
-RUN echo "markdown-mermaid-saml-solved image complete"
-
-# Build the final image based on the IMAGE_TYPE specified.
-FROM ${IMAGE_TYPE} AS final
-# Redeclare IMAGE_TYPE variable so it can be referenced inside this build
-# stage per
-# https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact
-ARG IMAGE_TYPE
-RUN echo "Used ${IMAGE_TYPE} image type"
+    && su -s /bin/bash -c 'gem install bundler' "${CONTAINER_APP_USERNAME}" \
+    && su -s /bin/bash -c '${CONTAINER_APP_ROOT}/app/bin/bundle install --gemfile=${PLUGINS_DIR}/discourse-saml/Gemfile --path=${PLUGINS_DIR}/discourse-saml/gems' "${CONTAINER_APP_USERNAME}" \
+    && ln -s "${PLUGINS_DIR}/discourse-saml/gems/ruby/"* "${PLUGINS_DIR}/discourse-saml/gems/" \
+    && su -s /bin/bash -c '${CONTAINER_APP_ROOT}/app/bin/bundle install' "${CONTAINER_APP_USERNAME}"
```

Comment:
Does this help with reducing the startup of the discourse charm? I have been using the charm and after the pod starts it has to run another ~5 minutes of scripts for it to be ready to serve. Not sure if it is possible to run those scripts in the Dockerfile rather than after startup

Reply:
I plan to move some things in the script to the dockerfile. The leader election mechanism will change too, so that should also speed up things.

Reply:
Some of the start up time may be unavoidable as I understand it. Discourse needs to recompile all assets if any of its configuration variables have changed I believe, which obviously they will at run time versus build time.

Reply:
That's my understanding too, but I think there's some for significant improvement.

Reply:
:( Hopefully we can at least do something to improve startup time

Reply:
That would be nice :)